sentry-apache-http-client-5/src/main/java/io/sentry/transport/apache/ApacheHttpClientTransport.java

```diff
@@ -0,0 +1,147 @@
+import java.util.zip.GZIPOutputStream;
+import org.apache.hc.client5.http.async.methods.SimpleHttpRequest;
+import org.apache.hc.client5.http.async.methods.SimpleHttpRequests;
+import org.apache.hc.client5.http.async.methods.SimpleHttpResponse;
+import org.apache.hc.client5.http.impl.async.CloseableHttpAsyncClient;
+import org.apache.hc.core5.concurrent.FutureCallback;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.io.CloseMode;
+import org.apache.hc.core5.util.TimeValue;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * {@link ITransport} implementation that executes request asynchronously in a non-blocking manner
+ * using Apache Http Client 5.
+ */
+public final class ApacheHttpClientTransport implements ITransport {
+  private final @NotNull SentryOptions options;
+  private final @NotNull RequestDetails requestDetails;
+  private final @NotNull CloseableHttpAsyncClient httpclient;
+  private final @NotNull RateLimiter rateLimiter;
+  private final @NotNull AtomicInteger currentlyRunning;
+
+  public ApacheHttpClientTransport(
+      final @NotNull SentryOptions options,
+      final @NotNull RequestDetails requestDetails,
+      final @NotNull CloseableHttpAsyncClient httpclient,
+      final @NotNull RateLimiter rateLimiter) {
+    this(options, requestDetails, httpclient, rateLimiter, new AtomicInteger());
+  }
+
+  ApacheHttpClientTransport(
+      final @NotNull SentryOptions options,
+      final @NotNull RequestDetails requestDetails,
+      final @NotNull CloseableHttpAsyncClient httpclient,
+      final @NotNull RateLimiter rateLimiter,
+      final @NotNull AtomicInteger currentlyRunning) {
+    this.options = Objects.requireNonNull(options, "options is required");
+    this.requestDetails = Objects.requireNonNull(requestDetails, "requestDetails is required");
+    this.httpclient = Objects.requireNonNull(httpclient, "httpclient is required");
+    this.rateLimiter = Objects.requireNonNull(rateLimiter, "rateLimiter is required");
+    this.currentlyRunning =
+        Objects.requireNonNull(currentlyRunning, "currentlyRunning is required");
+    this.httpclient.start();
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void send(SentryEnvelope envelope, Object hint) throws IOException {
+    if (isSchedulingAllowed()) {
+      final SentryEnvelope filteredEnvelope = rateLimiter.filter(envelope, hint);
+
+      if (filteredEnvelope != null) {
+        currentlyRunning.incrementAndGet();
+
+        try (final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            final GZIPOutputStream gzip = new GZIPOutputStream(outputStream)) {
+          options.getSerializer().serialize(filteredEnvelope, gzip);
+
+          final SimpleHttpRequest request =
+              SimpleHttpRequests.post(requestDetails.getUrl().toString());
+          request.setBody(outputStream.toByteArray(), ContentType.APPLICATION_JSON);
```

Comment:
> Reading the stream into a bytearray sounds suboptimal.

As discussed it will be changed in the future. There is no straightforward way to write to output of the Apache client's request.